packages/custom-elements/tests/html/HTMLElement/constructor.html

```diff
@@ -72,25 +62,235 @@
+        constructor() {
+          if (createSelf) {
+            createSelf = false;
+            // When this inner constructor calls into HTMLElement, it will pop
+            // the element that the polyfill is attempting to upgrade with the
+            // outer call. That element will never have had its state set,
+            // causing `createAnElement` to report an error.
+            document.createElement(localName);
+          }
+
+          // Then, the outer constructor's call into HTMLElement sees the
+          // AlreadyConstructedMarker on the stack instead of the expected
+          // element (because the inner call removed it) and also reports an
+          // error.
+          super();
+        }
+      }
+
+      const errors = [];
+      catchReportedErrors(() => {
+        assert.doesNotThrow(() => {
+          customElements.define(localName, NonConformantCustomElement);
+        });
+      }, (e) => {
+        e.preventDefault();
+        e.stopImmediatePropagation();
+        errors.push(e);
+      });
+      assert.equal(errors.length, 2);
+
+      document.body.removeChild(element);
+    });
+
+  test('Non-conformant custom element constructors that return an element ' +
+    'other than the one being constructed DO report an error when ' +
+    'constructed by `Document#createElement`, but not by direct construction.',
+    function() {
+      class NonConformantCustomElement extends HTMLElement {
+        constructor() {
+          return document.createElement('div');
+        }
+      }
+      customElements.define(localName, NonConformantCustomElement);
+
+      const errors = [];
+      catchReportedErrors(() => {
+        assert.doesNotThrow(function() {
+          document.createElement(localName);
+        });
+      }, (e) => {
+        e.preventDefault();
+        e.stopImmediatePropagation();
+        errors.push(e);
+      });
+      assert.equal(errors.length, 1);
+
+      assert.doesNotThrow(function() {
+        new NonConformantCustomElement();
+      });
+    });
+
+  test('Non-conformant custom element constructors that return an element ' +
+    'other than the one being constructed DO report an error when upgraded.',
+    function() {
+      const element = document.createElement(localName);
+      document.body.appendChild(element);
+
+      class NonConformantCustomElement extends HTMLElement {
+        constructor() {
+          return document.createElement('div');
+        }
+      }
+
+      const errors = [];
+      catchReportedErrors(() => {
+        assert.doesNotThrow(function() {
+          customElements.define(localName, NonConformantCustomElement);
+        });
+      }, (e) => {
+        e.preventDefault();
+        e.stopImmediatePropagation();
+        errors.push(e);
+      });
+      assert.equal(errors.length, 1);
+
+      document.body.removeChild(element);
+    });
+
+  test('Constructor errors do not halt upgrades in a tree.', function() {
```

Comment:
Let's add a test that shows errors in callbacks do not halt upgrades.

Reply:
Ok, added.